Merge llvm 401

juj opened the pull request.

juj commented:
This updates our LLVM from upstream LLVM 4.0 to LLVM 4.0.1, i.e. up to these commits:

http://llvm.org/viewvc/llvm-project/llvm/tags/RELEASE_401/final/ and
http://llvm.org/viewvc/llvm-project/cfe/tags/RELEASE_401/final/

The two last commits are the merge and API update, i.e.
https://github.com/kripken/emscripten-fastcomp/pull/191/commits/82738ace9886c062c5808bd012c66005c9a88251 does the merge, conflicted files:

```
# Conflicts:
#	include/llvm/Analysis/AssumptionCache.h
#	include/llvm/InitializePasses.h
#	lib/Analysis/AssumptionCache.cpp
#	lib/Transforms/Scalar/NewGVN.cpp
#	test/Transforms/NewGVN/pr31613.ll
#	tools/llvm-xray/xray-account.cc
```

some of which were pretty odd conflicts that I don't understand where they came from. I basically used the upstream version wherever there was no Emscripten LOCALMOD comment marked down.

The more important bit to review is then the update from LLVM 4.0 API to LLVM 4.0.1: https://github.com/kripken/emscripten-fastcomp/pull/191/commits/d638ded22add6c94e48b5fe96b7bc851f3f054f5.

The current status is that this is not quite ready to merge. The branch builds, but crashes on the ExpandVarArgs.cpp pass, in handling `llvm.lifetime.start` intrinsic, which is starting to get beyond my understanding. The crash comes from `auto LifetimeStart = Intrinsic::getDeclaration(M, Intrinsic::lifetime_start);` in ExpandVarArgs.cpp:228, where it looks like it's seeing the second parameter incorrect, in https://llvm.org/docs/LangRef.html#llvm-lifetime-start-intrinsic it says llvm.lifetime.start should be

`declare void @llvm.lifetime.start(i64 <size>, i8* nocapture <ptr>)`

however in debugger, it is seen as

![llvm lifetime start](https://user-images.githubusercontent.com/225351/28389914-ac05ebfe-6ce0-11e7-8687-aa896249648d.png)

but instead I would have expected the second parameter to say "Pointer" or similar. This discrepancy then leads to dereferencing an empty array. @kripken: does this look any familiar to you? It is possible I did the API update wrong with `AttributeSet` and `AttributeList`, I'm not quite sure about those.

juj commented:
I was able to resolve the above `llvm.lifetime.start` related issue, it looks like its signature has changed. The update is in this commit https://github.com/kripken/emscripten-fastcomp/pull/191/commits/ba3319708a09b41f4bf167b1f8ca492246facd64. Still running tests..

kripken commented:
Running tests over here, I see some failures:

 * `test_isdigit_l, test_istream, test_std_cout_new`, etc., invalid function pointer
 * `test_rounding`
 * `test_sqlite (sscanf?)` (shows `fastcomp/include/llvm/IR/Type.h:368: llvm::Type* llvm::Type::getPointerElementType() const: Assertion `getTypeID() == PointerTyID' failed.`)
 * `test_sse2_full` (shows `warning: unresolved symbol: llvm_rint_f64`, possibly that needs to be implemented)

and in `other`:
````
ERROR: test_cmake (test_other.other)
ERROR: test_create_readonly (test_other.other)
ERROR: test_eval_ctors (test_other.other)
ERROR: test_llvm_lto (test_other.other)
ERROR: test_locale_wrong (test_other.other)
ERROR: test_merge_pair (test_other.other)
ERROR: test_only_force_stdlibs (test_other.other)
ERROR: test_only_force_stdlibs_2 (test_other.other)
ERROR: test_readdir_r_silly (test_other.other)
ERROR: test_scons (test_other.other)
ERROR: test_source_file_with_fixed_language_mode (test_other.other)
ERROR: test_strftime_zZ (test_other.other)
ERROR: test_truncate_from_0 (test_other.other)
ERROR: test_vorbis (test_other.other)
FAIL: test_embind (test_other.other)
FAIL: test_emcc (test_other.other)
FAIL: test_emterpreter_file_suggestion (test_other.other)
FAIL: test_llvm_lit (test_other.other)
FAIL: test_nosplit (test_other.other)
````

kripken commented:
One issue was the lack of `llvm.rint.*` support. I added that in `next-merge2` branches in emscripten and fastcomp.

dylanmckay commented on lib/Target/JSBackend/AllocaManager.cpp:
```diff
@@ -541,8 +541,8 @@ void AllocaManager::analyze(const Function &Func, const DataLayout &Layout,
   const Module *M = F->getParent();
-  LifetimeStart = M->getFunction(Intrinsic::getName(Intrinsic::lifetime_start));
-  LifetimeEnd = M->getFunction(Intrinsic::getName(Intrinsic::lifetime_end));
+  LifetimeStart = M->getFunction("llvm.lifetime.start");
```
With this PR, the AllocaManager never actually does anything.

In `test/CodeGen/JS/allocamanger.ll`, `llvm.lifetime.start` and `llvm.lifetime.end` are defined and called, but for some reason, `LifetimeStart` and `LifetimeEnd` are both `nullptr` in my debugger.

This causes

```cpp
  552   if (PerformColoring &&
  553   ¦ ¦ ((LifetimeStart && !LifetimeStart->use_empty()) ||
  554   ¦ ¦ ¦(LifetimeEnd   && !LifetimeEnd->use_empty()))) {
```

To never evaluate to true.

juj replied:
Good find - I'll have to find time to get back to this and resolve the remaining issues, but unfortunately I will be travelling for the next three weeks, so I am not sure right now how many cycles I will have. Was your LLVM 5.0 merge branch free from this issue, or e.g. did you have a better fix there?

dylanmckay replied:
I believe my PR has the same issue. With that branch, I got everything compiling successfully and debugged the `allocamanager.ll` failure, but I haven't ran any tests other than that.

I've left my 5.0 PR on hold and will focus on getting this PR fixed if you don't have the time. I also don't have much time at the moment, so it will likely be a slowish process

juj replied:
Currently all tests pass, and I did not run into any problems around this region of code. So I wonder if that might be something that happens in the later LLVM 5 update(?)

juj commented:
Updated this branch and fixed some issues. Now I've got quite far, and I think there's only a couple of issues remaining.

One bigger trouble I see is something I could use @kripken's help with. I see that with these three branches applied:
 - https://github.com/juj/emscripten/tree/merge_llvm_401
 - https://github.com/juj/emscripten-fastcomp/tree/merge_llvm_401
 - https://github.com/juj/emscripten-fastcomp-clang/tree/merge_clang_401

the flag `-s DISABLE_EXCEPTION_CATCHING=1` is not working. This can be reproduced with

`exception.cpp`:
```
int main()
{
  try { throw 1; } catch(...) {}
}
```

`em++ exception.cpp -o a.js -s DISABLE_EXCEPTION_CATCHING=1`

This fails with

```
C:\code\emsdk\emscripten\incoming>em++ exception.cpp -o a.js -s DISABLE_EXCEPTION_CATCHING=1
Value:   %9 = landingpad { i8*, i32 }
          cleanup
LLVM ERROR: Unrecognized struct value
Traceback (most recent call last):
  File "C:\code\emsdk\emscripten\incoming\\em++", line 16, in <module>
    emcc.run()
  File "C:\code\emsdk\emscripten\incoming\emcc.py", line 1557, in run
    final = shared.Building.emscripten(final, append_ext=False, extra_args=extra_args)
  File "C:\code\emsdk\emscripten\incoming\tools\shared.py", line 2021, in emscripten
    call_emscripten(cmdline)
  File "C:\code\emsdk\emscripten\incoming\emscripten.py", line 2246, in _main
    temp_files.run_and_clean(lambda: main(
  File "C:\code\emsdk\emscripten\incoming\tools\tempfiles.py", line 79, in run_and_clean
    return func()
  File "C:\code\emsdk\emscripten\incoming\emscripten.py", line 2251, in <lambda>
    DEBUG=DEBUG,
  File "C:\code\emsdk\emscripten\incoming\emscripten.py", line 2151, in main
    temp_files=temp_files, DEBUG=DEBUG)
  File "C:\code\emsdk\emscripten\incoming\emscripten.py", line 93, in emscript
    backend_output = compile_js(infile, settings, temp_files, DEBUG)
  File "C:\code\emsdk\emscripten\incoming\emscripten.py", line 122, in compile_js
    shared.jsrun.timeout_run(subprocess.Popen(backend_args, stdout=subprocess.PIPE), note_args=backend_args)
  File "C:\code\emsdk\emscripten\incoming\tools\jsrun.py", line 20, in timeout_run
    raise Exception('Subprocess "' + ' '.join(note_args) + '" failed with exit code ' + str(proc.returncode) + '!')
Exception: Subprocess "C:/code/emsdk/clang/fastcomp/build_incoming_vs2015_64/Release/bin\llc c:\users\clb\appdata\local\temp\tmpguzw5p\a.bc -march=js -filetype=asm -o c:\users\clb\appdata\local\temp\tmplshuov.4.js -emscripten-stack-size=5242880 -O0 -emscripten-assertions=1 -emscripten-no-aliasing-function-pointers -emscripten-global-base=8" failed with exit code 1!
```

When I cross-reference the `.ll` files before and after the merge, I see that they are practically identical for `exception.cpp` above:

before: [exception_cpp_before_merge_ll.txt](https://github.com/kripken/emscripten-fastcomp/files/1418272/exception_cpp_before_merge_ll.txt)

after: [exception_cpp_after_merge_ll.txt](https://github.com/kripken/emscripten-fastcomp/files/1418276/exception_cpp_after_merge_ll.txt)

so LLVM bitcode generation has not changed in that part at least. The failure comes from NaCl's `ExpandStructRegs.cpp` pass, I think when operating on the following:

```
  %1 = landingpad { i8*, i32 }
          catch i8* null
  %2 = extractvalue { i8*, i32 } %1, 0
```

If I step through code in the pre-merge version, the `ExpandStructRegs` pass is never called for the `landingpad` instruction, and when I trace the set of instructions that NaCl's `ExpandStructRegs` pass does go through, with

```
static bool ExpandExtractValues(Function &Func, bool Finalize) {
  bool Changed = false;

  SmallVector<Instruction *, 10> ToErase;
  // Expand out all the extractvalue instructions.  Also collect up
  // the insertvalue instructions for later deletion so that we do not
  // need to make extra passes across the whole function.

  for (auto &BB : Func) {
      DEBUG(dbgs() << "BB " << Func << "\n");       // ********
    for (BasicBlock::iterator Iter = BB.begin(), E = BB.end(); Iter != E;) {
      Instruction *Inst = &*Iter++;
      DEBUG(dbgs() << "Inst " << *Inst<< "\n"); // ********
      if (ExtractValueInst *EV = dyn_cast<ExtractValueInst>(Inst)) {
        DEBUG(dbgs() << "EVInst " << *EV << "\n"); // ********
        Changed |= ExpandExtractValue(EV, &ToErase);
      } else if (isa<InsertValueInst>(Inst)) {
        ToErase.push_back(Inst);
        Changed = true;
      }
    }
  }
```

I see that before merge, the function call `ExpandExtractValue()` is not reached, because of the function that is being processed looks like

```
; Function Attrs: noinline norecurse
define i32 @main() #0 personality i8* bitcast (i32 (i8*)* @__gxx_personality_v0 to i8*) {
entry:
  %exn.slot = alloca i8*
  %ehselector.slot = alloca i32
  %exception = call i8* @__cxa_allocate_exception(i32 4) #1
  %0 = bitcast i8* %exception to i32*
  store i32 1, i32* %0, align 16
  call void @__cxa_throw(i8* %exception, i8* bitcast (i8** @_ZTIi to i8*), i8* null) #2
  unreachable
}
```

whereas after the merge, the function call that is being processed looks like

```
; Function Attrs: noinline norecurse optnone
define i32 @main() #0 personality i8* bitcast (i32 (...)* @__gxx_personality_v0 to i8*) {
  %1 = alloca i8*
  %2 = alloca i32
  %3 = call i8* @__cxa_allocate_exception(i32 4) #1
  %4 = bitcast i8* %3 to i32*
  store i32 1, i32* %4, align 16
  invoke void @__cxa_throw(i8* %3, i8* bitcast (i8** @_ZTIi to i8*), i8* null) #2
          to label %13 unwind label %5

; <label>:5:                                      ; preds = %0
  %6 = landingpad { i8*, i32 }
          catch i8* null
  %7 = extractvalue { i8*, i32 } %6, 0
  store i8* %7, i8** %1, align 4
  %8 = extractvalue { i8*, i32 } %6, 1
  store i32 %8, i32* %2, align 4
  br label %9

; <label>:9:                                      ; preds = %5
  %10 = load i8*, i8** %1, align 4
  %11 = call i8* @__cxa_begin_catch(i8* %10) #1
  call void @__cxa_end_catch()
  br label %12

; <label>:12:                                     ; preds = %9
  ret i32 0

; <label>:13:                                     ; preds = %0
  unreachable
}
```

The difference in these two is that it looks like `__cxa_throw` is no longer immediately followed by a `uncreachable`, which leads to processing the `landingpad` instruction. I wonder if the problem is that 

a) somehow these `unreachable`s got lost during the merge (and we should find to restore them back?), or

b) some LLVM behavior changed between LLVM 4.0.0 to LLVM 4.0.1 and the new code after the `__cxa_throw` is now desired (and we should find to make a way for `ExpandStructRegs` to be able to handle the `landingpad` instruction?)

I tried going down the path b) and fix `ExpandStructRegs` to be able to process an `extractvalue` to a `landingpad` instruction, but I'm afraid I don't really know what I'm doing there, and just ended up with crashes. @kripken : does this look like something that would be familiar to you? Help would be much appreciated!

juj commented:
The above failure ends up erroring the following tests in the test suite:

```
test_dylink_dynamic_cast (test_core.default)
test_dylink_iostream (test_core.default)
test_dylink_syslibs (test_core.default)
test_exceptions (test_core.default)
test_exceptions_white_list (test_core.default)
test_fs_writeFile (test_core.default)
test_longjmp_throw (test_core.default)
```

Curiously only the default test mode is affected, in asm1 and other modes the tests pass ok, so presumably the extra code does get DCEd away in those modes.

juj commented:
Finished going through the core, other and browser suites, and it looks like we are looking at only four remaining issues:
 - #202: Test `other.test_nosplit` failure, the compiler generates slightly worse code to fail this test,
 - #203: `-s DISABLE_EXCEPTION_CATCHING=1` not working if compiling with `-O0`,
 - #204: Test `other.test_emcc` fails since `--llvm-opts 1` is no longer enough(?) to trigger the `printf`->`puts` optimization
 - https://github.com/kripken/emscripten/issues/5716: emterpreter handles rounding mode of `round()` incorrectly

All other tests are passing ok.

kripken commented:
About the exception issue: what does the bitcode look like without `-s DISABLE_EXCEPTION_CATCHING=1`, i.e., when we do want exceptions? It should have landingpads in that case, so the code must be handling them somehow. So first question is, why can it handle landingpads in that case but not the other.

My guess is that when exceptions are enabled we turn those landingpads into something simpler before ExpandStructRegs even sees it (see the passes added in JSBackend.cpp near the end). It's probably the exception-handling pass, `createLowerEmExceptionsPass`. So in theory we could run that pass even when exceptions are disabled, to prevent the compiler crash.

But the bigger issue is why llvm is now emitting much worse code here, it's like it isn't aware exceptions are disabled, which they should be with `-fno-exceptions` or whatever that flag is. So maybe in the merge something broke with passing that flag? Or how it is used?

juj commented:
> My guess is that when exceptions are enabled we turn those landingpads into something simpler before ExpandStructRegs even sees it

This ended up being spot on, commented in https://github.com/kripken/emscripten-fastcomp/issues/203#issuecomment-340006394.

That resolves #203 and #204. The issues #202 and https://github.com/kripken/emscripten/issues/5716 are the only ones then that remain.

kripken commented:
Planning note: I think we should merge incoming to master before this lands. It's something we really need to do anyhow, and doing it before a big LLVM merge is a good idea to minimize risk. Thoughts?

dschuff commented:
I think that's a good idea.

kripken commented:
On second thought, @juj reports that Unity hasn't been using incoming due to an LLVM but which is fixed in 4.0.1. So we should probably merge 4.0.1 first, then see about merging to master (so we don't merge that LLVM bug into master).

juj commented:
All PR branches are now updated:
 - https://github.com/kripken/emscripten/pull/5400
 - https://github.com/kripken/emscripten-fastcomp/pull/191
 - https://github.com/kripken/emscripten-fastcomp-clang/pull/18

and all tests should pass. Ran all core suite variants and `other` suite on both Linux and Windows, and I don't see any failures. @kripken: these should now be all ready to merge - tag both before and after?

kripken commented:
I'll start to review now, great.

juj commented:
In particular check out the three commits shown above in this GitHub view, those are my changes, the first one is a "stock" merge with only merge conflicts resolved, but no other changes.

kripken commented on lib/Target/JSBackend/NaCl/ExpandVarArgs.cpp:
```diff
@@ -190,7 +190,7 @@ static bool ExpandVarArgCall(Module *M, InstType *Call, DataLayout *DL) {
   for (unsigned I = 0, E = FuncType->getNumParams(); I < E; ++I) {
     FixedArgs.push_back(Call->getArgOperand(I));
     // AttributeSets use 1-based indexing.
```
isn't this another case of the off-by-one issue that you just fixed in ExpandByVal?

juj replied:
That is a good question. I don't actually understand where that off by one issue comes from, and what exactly changed so that I had to do that change in ExpandByVal. I did that change in ExpandByVal because I saw in test code that the expansion by value was getting applied to parameters "off by one" in the generated JS code. It may be the case that this + 1 needs to be removed, or then not. In the current branch all tests pass, so I don't want to just remove the + 1 here without understanding.

kripken replied:
This gives me low confidence in the +1 removal change. I think we need to understand this better.

How about asking for help from Rust people, which they offered in the other PR? Many of them are LLVM experts. Perhaps they can give a quick and clear response on changes to the Attribute API changes in 4.0.1.

juj replied:
Agreed, I am reading the changes, but I don't know who would be a good person to review this.

kripken replied:
Can ask where the Rust people offered to help, in the 5.0 PR I think it was.

juj replied:
Ok, commented in https://github.com/kripken/emscripten-fastcomp/pull/199#issuecomment-341769802

juj replied:
Upon further inspection, it looks like the answer to this is yes.

kripken commented on lib/Target/JSBackend/NaCl/SimplifyStructRegSignatures.cpp:
```diff
@@ -219,42 +219,31 @@ static AttributeSet CopyRetAttributes(LLVMContext &C, const DataLayout &DL,
                                       const unsigned RetIndex) {
   const unsigned NewIndex = RetIndex + OldArg + 1;
   if (!NewArgTy) {
     const unsigned OldIndex = OldArg + 1;
```
Also here?

juj replied:
Yes.

juj commented:
Also ran the `browser` suite on Windows, which passes.

juj commented:
Now I understand where the discrepancy comes from. The `AttributeList` object still uses 1-based indexing in the sense that at index 0 is the return value, and at index 1...N are the parameters to the function; but what seems to be the breaking change here is that as `AttributeSet` got refactored to `AttributeList`, the input parameter to the function `AttributeList::getParamAttributes(int Index)` was changed; it used to be that `AttributeSet::getParamAttributes(1)` returned the attributes of the first function parameter, but now what happens is that one needs to call `AttributeList::getParamAttributes(0)` to get the attributes of the first parameter. For other API entry points the essential 1-based indexing still exists, e.g. `AttributeSet::getAttributes(i)` is still the same as `AttributeList::getAttributes(i)`, so that function did not get shifted.

Rerunning all tests after the above changes, and I'll push an updated version shortly.

juj commented:
Updated with the corrected `+ 1` adjustments in above commit. All tests pass in core and `other` suites, so this is good for a second look now?

kripken commented:
Thanks! Looks good. It does appear that `getParamAttributes` now adds `FirstArgIndex` (which is equal to one).

kripken commented on lib/Target/JSBackend/NaCl/ExpandI64.cpp:
```diff
@@ -230,12 +230,12 @@ static Function *RecreateFunctionLegalized(Function *F, FunctionType *NewType) {
       continue;
     }
-    if (!Attrs.hasAttributes(i)) continue;
+    if (!Attrs.hasAttributes(i+1)) continue;
```
this can be `hasParamAttribute(i)`

juj replied:
check

juj commented:
Ok, did a new pass on this, bringing all three branches up to date with latest, and reran default and other suites, which still pass, except for `other.test_llvm_lit` failing. It would be great if you have a sec to look at this @kripken . What's happening is it looks like perhaps some DCE is not executed? The failure reads

```
test_llvm_lit (test_other.other) ... -- Testing: 29 tests, 29 threads --
******************** TEST 'LLVM :: CodeGen/JS/allocamanager-phis.ll' FAILED ********************
Script:
--
/home/clb/emsdk/clang/fastcomp/build_incoming_64/./bin/llc < /home/clb/emsdk/clang/fastcomp/src/test/CodeGen/JS/allocamanager-phis.ll | /home/clb/emsdk/clang/fastcomp/build_incoming_64/./bin/FileCheck /home/clb/emsdk/clang/fastcomp/src/test/CodeGen/JS/allocamanager-phis.ll
--
Exit Code: 1

Command Output (stderr):
--
/home/clb/emsdk/clang/fastcomp/src/test/CodeGen/JS/allocamanager-phis.ll:9:10: error: expected string not found in input
; CHECK: $vararg_buffer3 = sp;
         ^
<stdin>:2:1: note: scanning from here
// EMSCRIPTEN_START_FUNCTIONS
^
<stdin>:8:2: note: possible intended match here
 $vararg_buffer3 = sp + 16|0;
 ^

--

********************
******************** TEST 'LLVM :: CodeGen/JS/allocamanager.ll' FAILED ********************
Script:
--
/home/clb/emsdk/clang/fastcomp/build_incoming_64/./bin/llc < /home/clb/emsdk/clang/fastcomp/src/test/CodeGen/JS/allocamanager.ll | /home/clb/emsdk/clang/fastcomp/build_incoming_64/./bin/FileCheck /home/clb/emsdk/clang/fastcomp/src/test/CodeGen/JS/allocamanager.ll
--
Exit Code: 1

Command Output (stderr):
--
/home/clb/emsdk/clang/fastcomp/src/test/CodeGen/JS/allocamanager.ll:20:10: error: expected string not found in input
; CHECK: STACKTOP = STACKTOP + 128|0;
         ^
<stdin>:4:2: note: scanning from here
 $argc = $argc|0;
 ^
<stdin>:8:2: note: possible intended match here
 STACKTOP = STACKTOP + 272|0;
 ^

--

********************
********************
Failing Tests (2):
    LLVM :: CodeGen/JS/allocamanager-phis.ll
    LLVM :: CodeGen/JS/allocamanager.ll

  Expected Passes    : 27
  Unexpected Failures: 2
FAIL

======================================================================
FAIL: test_llvm_lit (test_other.other)
----------------------------------------------------------------------
```

I can make these two tests pass with

```diff
diff --git a/test/CodeGen/JS/allocamanager-phis.ll b/test/CodeGen/JS/allocamanager-phis.ll
index c04a212..71d5737 100644
--- a/test/CodeGen/JS/allocamanager-phis.ll
+++ b/test/CodeGen/JS/allocamanager-phis.ll
@@ -6,9 +6,9 @@
 ; an overlapping lifetime with l_766.i which is only visible if we can
 ; see through phis.

-; CHECK: $vararg_buffer3 = sp;
-; CHECK: $l_1565$i = sp + 16|0;
-; CHECK: $l_766$i = sp + 12|0;
+; CHECK: $vararg_buffer3 = sp + 16|0;
+; CHECK: $l_1565$i = sp + 36|0;
+; CHECK: $l_767$i = sp + 32|0;

 target datalayout = "e-p:32:32-i64:64-v128:32:128-n32-S128"
 target triple = "asmjs-unknown-emscripten"
diff --git a/test/CodeGen/JS/allocamanager.ll b/test/CodeGen/JS/allocamanager.ll
index 19f1ca7..d8346ab 100644
--- a/test/CodeGen/JS/allocamanager.ll
+++ b/test/CodeGen/JS/allocamanager.ll
@@ -16,11 +16,9 @@ target triple = "asmjs-unknown-emscripten"
 @.str3 = private unnamed_addr constant [43 x i8] c"another message from the program: \22%s\22...\0A\00", align 1

 ; CHECK: function _foo($argc,$argv) {
-; CHECK-NOT: cupcake
-; CHECK: STACKTOP = STACKTOP + 128|0;
+; CHECK: STACKTOP = STACKTOP + 272|0;
 ; CHECK-NEXT: vararg_buffer0 =
-; CHECK-NEXT: $muffin =
-; CHECK-NOT: cupcake
+; CHECK: $muffin =
 ; CHECK: }

 ; Function Attrs: nounwind
@@ -73,11 +71,8 @@ entry:
 }

 ; CHECK: function _bar($argc,$argv) {
-; CHECK-NOT: cupcake
-; CHECK: STACKTOP = STACKTOP + 128|0;
+; CHECK: STACKTOP = STACKTOP + 272|0;
 ; CHECK-NEXT: vararg_buffer0 =
-; CHECK-NEXT: $muffin =
-; CHECK-NOT: cupcake
 ; CHECK: }

 ; Function Attrs: nounwind
```

but that doesn't obviously look good. For the first test `allocamanager-phis.ll` I'm not sure if it's natural that there could be one more variable on the stack, bumping each stack pointer by 16 bytes. But for the second test it looks like the point of the test is to verify that `cupcake` gets optimized away, although it doesn't. Does this look like something sensible?

Apart from this failure, all other tests are passing ok.

kripken commented:
I'm not that familiar with the AllocaManager code. @sunfishcode, you wrote it, any thoughts on what @juj is seeing above?

juj commented:
The above commit fixes the AllocaManager test failures. Tested that the core and other suites now pass, so this is good to land based on everything I'm seeing.

@kripken : have we ever updated LLVM without also updating Emscripten minor version? I.e. should we do a 1.38.0 out of this, or a 1.37.30?

juj commented:
> Also need to update the llvm_lifetime_* names in CallHandlers.h

In that file, I see I went with having both variants:

```c++
DEF_CALL_HANDLER(llvm_lifetime_start, {
  return "";
})

DEF_CALL_HANDLER(llvm_lifetime_end, {
  return "";
})

DEF_CALL_HANDLER(llvm_lifetime_start_p0i8, {
  return "";
})

DEF_CALL_HANDLER(llvm_lifetime_end_p0i8, {
  return "";
})

DEF_CALL_HANDLER(llvm_invariant_start, {
  return "";
})

DEF_CALL_HANDLER(llvm_invariant_end, {
  return "";
})

DEF_CALL_HANDLER(llvm_invariant_start_p0i8, {
  return "";
})

DEF_CALL_HANDLER(llvm_invariant_end_p0i8, {
  return "";
})
```

this allows building code coming from old .bc files, I think.

kripken commented:
I would also check `sanity` and `browser`.

What we've done in the past is tag a minor version before and after. So `1.37.30` and `1.37.31`. (To tag `1.38.0` we should merge to master, which we should be more careful about, but we should eventually do that too.)

juj commented:
Checked through `browser` and `sanity`. Browser runs fine without issues, but `sanity` is a bit tricky suite, because it expects an environment that emsdk-based installations are not compatible with, e.g. to run `sanity`, `EMCC_CONFIG` can't be defined, and a prebuilt Binaryen or asm.js optimizer can't be present. I could not find any failures in `sanity` that looked like to be caused by this PR though, so would be confident to merge this now. Do you want to do further tests before so?

kripken commented:
That's ok, you can merge this. I'll run the sanity tests later, almost certainly any issue there is unrelated to LLVM changes.

juj commented:
Perfect, tagging and merging and tagging now..
